### PR TITLE
SilverStripe 4 compatibility

### PR DESCRIPTION
**Pull request by robbieaverill**
Updated for compatibility with SilverStripe 4 in its current state. Brief summary of changes:

* Composer: PSR-4 autoloading compatibility
* Structure: "code" -> "src", namespacing added and implemented
* Travis: config updated for SS4 (PHP 7.x is no longer allowed to fail)
* PSR-2: some style formatting changes
* Documentation: updated to include namespaced class names and imported in DataObject examples

### CODE REVIEW

**Comment by dhensby** [comment]
`master` is not aliased as `2.x-dev` which will be 4.x compat. Please update `.travis.yml` and `composer.json` constraints

**Comment by robbieaverill** [comment]
Branch rebased against the latest `master` with @dhensby's latest alias change in it.

File: .travis.yml [hunk 1]
```diff
@@ -4,26 +4,18 @@ sudo: false
...
-env:
-  - DB=MYSQL CORE_RELEASE=3.2
-
 matrix:
   include:
+    - php: 5.5
+      env: DB=MYSQL CORE_RELEASE=master
     - php: 5.6
-      env: DB=MYSQL CORE_RELEASE=3
+      env: DB=MYSQL CORE_RELEASE=master
```
Comment (dhensby) [inline .travis.yml]: I think using `CORE_RELEASE=4` will be better for "future proofing" the tests

Comment (robbieaverill) [inline .travis.yml]: Updated

**Comment by tractorcow** [comment]
Creating a new blog post redirects me to a strange url `http://ss40test.loc/http://ss40test.loc/admin/pages/edit/show/14`. Remove the $record->absoluteBaseURL() from GridFieldSiteTreeAddNewButton().

```php
public function handleAction(GridField $gridField, $actionName, $arguments, $data)
    {
        if ($actionName !== 'add') {
            return null;
        }
        $tmpData = json_decode($data['ChildPages']['GridState'], true);
        /** @skipUpgrade  */
        $tmpData = $tmpData['GridFieldSiteTreeAddNewButton'];

        $data = array(
            'ParentID' => $tmpData['currentPageID'],
            'PageType' => $tmpData['pageType']
        );

        $controller = Injector::inst()->create(CMSPageAddController::class);

        $form = $controller->AddForm();
        $form->loadDataFrom($data);

        $controller->doAdd($data, $form);
        $response = $controller->getResponseNegotiator()->getResponse();

        // Get the current record
        /** @var SiteTree $record */
        $record = SiteTree::get()->byId($controller->currentPageID());
        if ($record) {
            $response->redirect($record->CMSEditLink(), 301);
        }
        return $response;
    }
```

**Comment by robbieaverill** [comment]
Fixed- I'll test the three together more before I merge. Various things in the UI have changed recently and need an update. Thanks!

**Comment by dhensby** [comment]
Is this ready? would be good to get it merged so we can then get some other modules test-suites passing :)

**Comment by robbieaverill** [comment]
Yeah I reckon so

**Pull request merged**